libtransmission/peer-mgr.h

```diff
@@ -316,17 +351,48 @@ class tr_peer_info
         return ret;
     }
 
-private:
-    [[nodiscard]] constexpr tr_address const& addr() const noexcept
-    {
-        return socket_address_.address();
-    }
+    // ---
 
-    [[nodiscard]] constexpr auto port() const noexcept
+    // merge a peer info objects into ourself
+    void merge(tr_peer_info const& that) noexcept
     {
-        return socket_address_.port();
+        connection_attempted_at_ = std::max(connection_attempted_at_, that.connection_attempted_at_);
+        connection_changed_at_ = std::max(connection_changed_at_, that.connection_changed_at_);
+        piece_data_at_ = std::max(piece_data_at_, that.piece_data_at_);
+
+        /* no need to merge blocklist since it gets updated elsewhere */
+
+        {
+            auto const conn_this = !is_connectable_ || *is_connectable_;
+            auto const conn_that = !that.is_connectable_ || *that.is_connectable_;
+
+            if (conn_this != conn_that)
+            {
+                is_connectable_.reset();
+            }
+            else
+            {
+                set_connectable(conn_this && conn_that);
+            }
+        }
+
+        set_utp_supported(supports_utp() || that.supports_utp());
+
+        /* from_first_ should never be modified */
+        found_at(that.from_best());
+
+        /* num_consecutive_fails_ is already the latest */
+        set_pex_flags(that.pex_flags(), true);
```

Comment:
_(opinion)_  The `merge_unsupported` bool isn't used by anyone except other `tr_peer_info` methods (I think). Maybe take it out of the API by making the flag private?

```c++
public:
    void set_flags(flags)
    {
        set_flags(flags, false);
    }

private:
    void set_flags(flags, bool merge_unsupported)
    {
        ....
    }
````

Reply:
I think the code outside of the `tr_peer_info` class should be free to decide whether to overwrite the flags (`merge_unsupported = false`) or to `OR` the flags (`merge_unsupported = true`). It just happened that none of the code outside of the class methods set `merge_unsupported` to `true`.

For example, it can be argued both ways whether `ensure_peer_exists()` should overwrite or `OR` the flags. On one hand, you could argue that the most recently received peer info should be the most updated, so we should overwrite the flags. On the other hand, you can also argue that the sources where we retrieve the peer's info from might only be aware of a part of the peer's capabilities, so we should consolidate the titbits by `OR`ing them.